bigtable/client/row_reader.cc

```diff
@@ -0,0 +1,182 @@
+// Copyright 2017 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "bigtable/client/row_reader.h"
+
+#include <thread>
+
+namespace bigtable {
+inline namespace BIGTABLE_CLIENT_NS {}  // namespace BIGTABLE_CLIENT_NS
+
+// RowReader::iterator must satisfy the requirements of an InputIterator.
+namespace {
+using base_iterator = std::iterator<std::input_iterator_tag, Row>;
+
+static_assert(std::is_base_of<base_iterator, RowReader::iterator>::value,
+              "RowReader::iterator must be an InputIterator");
+static_assert(std::is_copy_constructible<RowReader::iterator>::value,
+              "RowReader::iterator must be CopyConstructible");
+static_assert(std::is_move_constructible<RowReader::iterator>::value,
+              "RowReader::iterator must be MoveConstructible");
+static_assert(std::is_copy_assignable<RowReader::iterator>::value,
+              "RowReader::iterator must be CopyAssignable");
+static_assert(std::is_move_assignable<RowReader::iterator>::value,
+              "RowReader::iterator must be MoveAssignable");
+static_assert(std::is_destructible<RowReader::iterator>::value,
+              "RowReader::iterator must be Destructible");
+
+// TODO(C++17): std::is_swappable
```

Comment:
This TODO was meant as a "when C++17 is available we can check that iterator works with `std::swap()` easily".  You can probably remove it, assuming the iterator does work with `std::swap`?